clash-ghc/src-ghc/Clash/GHC/PartialEval.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.GHC.PartialEval
+  ( ghcEvaluator
+  ) where
+
+import Control.Monad.State.Strict (State)
+import qualified Control.Monad.State.Strict as State
+import GHC.Integer.GMP.Internals (Integer(..))
+
+import Clash.Core.DataCon
+import Clash.Core.Evaluator.Models
+import Clash.Core.Evaluator.Semantics
+import Clash.Core.Literal
+import Clash.Core.Term
+import Clash.Core.Type
+
+ghcEvaluator :: Evaluator
+ghcEvaluator = Evaluator ghcEvaluate ghcQuote
+
+ghcEvaluate :: Term -> State Env Value
+ghcEvaluate term =
+  case term of
+    Var v -> evaluateVarWith ghcEvaluate v
+    Literal l -> evaluateLiteral l
+    Data dc -> evaluateDataWith ghcEvaluate dc
+    Prim p -> evaluatePrimWith ghcEvaluate ghcEvaluatePrim p
+    Lam x e -> evaluateLam x e
+    TyLam x e -> evaluateTyLam x e
+    App x y -> evaluateAppWith ghcEvaluate ghcEvaluatePrim ghcApply x (Left y)
+    TyApp x ty -> evaluateAppWith ghcEvaluate ghcEvaluatePrim ghcApply x (Right ty)
+    Letrec bs x -> evaluateLetrecWith ghcEvaluate bs x
+    Case x ty xs -> evaluateCaseWith ghcEvaluate ghcIsLiteral ghcIsData x ty xs
+    Cast x a b -> evaluateCastWith ghcEvaluate x a b
+    Tick ti x -> evaluateTickWith ghcEvaluate ti x
+
+-- TODO Implement evaluation for primitives and call here.
+--
+ghcEvaluatePrim :: PrimInfo -> [Either Value Type] -> State Env Value
+ghcEvaluatePrim p args = pure (VNeu $ NePrim p args)
+
+ghcIsLiteral :: Literal -> Pat -> Bool
+ghcIsLiteral l = \case
+  DataPat c [] [_]
+    |  IntegerLiteral i <- l
+    -> case i of
+         S# _  -> dcTag c == 1
+         Jp# _ -> dcTag c == 2
+         Jn# _ -> dcTag c == 3
+
+    |  NaturalLiteral i <- l
+    -> case i of
+         S# _  -> dcTag c == 1 && i >= 0
+         Jp# _ -> dcTag c == 2
+         Jn# _ -> False
+
```

Comment:
I'm not entirely sure on this one. The old evaluator didn't take these types into account here, but I suspect that something like, e.g. `Int`, would be turned into `Literal (IntLiteral n)` by the core2core pass, but maybe `Integer` and `Natural` might not be?

@christiaanb, do you know anything that might help?

Reply:
`IntLiteral i` corresponds to something of type `Int#` in Haskell. While both `IntegerLiteral i` and `App (Data S#) (IntLiteral i)` correspond to something of type `Integer` in Haskell. To expand further:

The following Haskell:
```haskell
case (x :: Int) of
  1  -> True
  _  -> False
```
is desugared to:
```haskell
case x of
  DataPat I# i# -> case i# of
   DefaultPat -> False
   LitPat 1# -> True
```
while the following Haskell:
```haskell
case (x :: Integer) of
  1 -> True
  _ -> False
```
is desugared to:
```haskell
case (x :: Integer) of
  DefaultPat -> False
  LitPat 1 -> True
```
but we can in Haskell also write:
```haskell
case (x :: Integer) of
  S# i# -> case i# of
    1# -> True
    _ -> False
  _ -> False
```
which is desugared to:
```haskell
case (x :: Integer) of
  DefaultPat -> False
  DataPat S# i# -> case i# of
    DefaultPat -> False
    LitPat 1# -> True
```
So only for `Integer` (and `Natural`) do we need to be able to match against `LitPat`and `DataPat`. While for the other data types the LitPat are all the unboxed versions.

Reply:
I'm not sure either.
I agreed that you should get the `IntLiteral` for 
```
case x of
  5 -> ....
```

But if you have:
```
y = case x of
  2 -> 3#
  I# x -> x
  _    -> error "should be unreachable"

x :: Int
x = fromIntegral (4 :: Signed 4) 
```
I'm afraid this code might get to the `| otherwise -> False`, conclude that the second alternative doesn't match and evaluate to the third alternative.

Reply:
That unreachable is removed by desugaring

Reply:
Ok, the third case doesn't exists.
But the point still stands, if it doesn't match the `I# x` pattern correctly it will run into this [error](https://github.com/clash-lang/clash-compiler/blob/cb90adaf8ece84d1c1d7951ca6e3647a90a4fd10/clash-lib/src/Clash/Core/Evaluator/Semantics.hs#L261)

Reply:
No, because Integer and Natural are the only special types. Only they can have a DataPat and LitPat match. All boxed types only have a DataPat match; and unboxed types only have a LitPat match.

Reply:
i.e. GHC will never generate LitPat matches for Boxed types, nor generate DataPat matches for unboxed types. If it does it’s an actual error.